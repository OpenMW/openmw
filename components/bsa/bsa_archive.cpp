/*
  OpenMW - The completely unofficial reimplementation of Morrowind
  Copyright (C) 2008-2010  Nicolay Korslund
  Email: < korslund@gmail.com >
  WWW: http://openmw.sourceforge.net/

  This file (cpp_bsaarchive.cpp) is part of the OpenMW package.

  OpenMW is distributed as free software: you can redistribute it
  and/or modify it under the terms of the GNU General Public License
  version 3, as published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  version 3 along with this program. If not, see
  http://www.gnu.org/licenses/ .

 */

#include "bsa_archive.hpp"

#include <OgreArchive.h>
#include <OgreArchiveFactory.h>
#include <OgreArchiveManager.h>
#include "bsa_file.hpp"
#include <libs/mangle/stream/clients/ogre_datastream.hpp>

using namespace Ogre;
using namespace Mangle::Stream;

/// An OGRE Archive wrapping a BSAFile archive
class BSAArchive : public Archive
{
private:
  BSAFile arc;

public:
  BSAArchive(const String& name, const String& data_dir)
             : Archive(name, "BSA")
  { arc.open(name, data_dir); }

  bool isCaseSensitive() const { return false; }

  // The archive is loaded in the constructor, and never unloaded.
  void load() {}
  void unload() {}

  DataStreamPtr open(const String& filename, bool recursive = true) const
  {
    // Get a non-const reference to arc. This is a hack and it's all
    // OGRE's fault. You should NOT expect an open() command not to
    // have any side effects on the archive, and hence this function
    // should not have been declared const in the first place.
    BSAFile *narc = (BSAFile*)&arc;

    // Open the file
    StreamPtr strm = narc->getFile(filename.c_str());

    // Wrap it into an Ogre::DataStream.
    return DataStreamPtr(new Mangle2OgreStream(strm));
  }

  // Check if the file exists.
  bool exists(const String& filename) { return arc.exists(filename.c_str()); }
  time_t getModifiedTime(const String&) { return 0; }

  // This is never called as far as I can see.
  StringVectorPtr list(bool recursive = true, bool dirs = false)
  {
    //std::cout << "list(" << recursive << ", " << dirs << ")\n";
    StringVectorPtr ptr = StringVectorPtr(new StringVector());
    return ptr;
  }

  // Also never called.
  FileInfoListPtr listFileInfo(bool recursive = true, bool dirs = false)
  {
    //std::cout << "listFileInfo(" << recursive << ", " << dirs << ")\n";
    FileInfoListPtr ptr = FileInfoListPtr(new FileInfoList());
    return ptr;
  }

  // After load() is called, find("*") is called once. It doesn't seem
  // to matter that we return an empty list, exists() gets called on
  // the correct files anyway.
  StringVectorPtr find(const String& pattern, bool recursive = true,
                       bool dirs = false)
  {
    //std::cout << "find(" << pattern << ", " << recursive
    //          << ", " << dirs << ")\n";
    StringVectorPtr ptr = StringVectorPtr(new StringVector());
    return ptr;
  }

  /* Gets called once for each of the ogre formats, *.program,
     *.material etc. We ignore all these.

     However, it's also called by MyGUI to find individual textures,
     and we can't ignore these since many of the GUI textures are
     located in BSAs. So instead we channel it through exists() and
     set up a single-element result list if the file is found.
  */
  FileInfoListPtr findFileInfo(const String& pattern, bool recursive = true,
                               bool dirs = false)
  {
    FileInfoListPtr ptr = FileInfoListPtr(new FileInfoList());

    // Check if the file exists (only works for single files - wild
    // cards and recursive search isn't implemented.)
    if(exists(pattern))
      {
        FileInfo fi;
        fi.archive = this;
        fi.filename = pattern;
        // It apparently doesn't matter that we return bogus
        // information
        fi.path = "";
        fi.compressedSize = fi.uncompressedSize = 0;

        ptr->push_back(fi);
      }

    return ptr;
  }
};

// An archive factory for BSA archives
class BSAArchiveFactory : public ArchiveFactory
{
private:
    String data;
public:
  BSAArchiveFactory(const String& idata) {data = idata;}
  const String& getType() const
  {
    static String name = "BSA";
    return name;
  }

  Archive *createInstance( const String& name )
  {
    return new BSAArchive(name, data);
  }

  void destroyInstance( Archive* arch) { delete arch; }
};

static bool init = false;
static void insertBSAFactory(const String& data_dir)
{
  if(!init)
    {
      ArchiveManager::getSingleton().addArchiveFactory( new BSAArchiveFactory(data_dir) );
      init = true;
    }
}

// The function below is the only publicly exposed part of this file

void addBSA(const std::string& name, const std::string& group, const std::string& data)
{
  insertBSAFactory(data);
  ResourceGroupManager::getSingleton().
    addResourceLocation(name, "BSA", group);
}
