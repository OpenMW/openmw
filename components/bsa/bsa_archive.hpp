/*
  OpenMW - The completely unofficial reimplementation of Morrowind
  Copyright (C) 2008-2010  Nicolay Korslund
  Email: < korslund@gmail.com >
  WWW: http://openmw.sourceforge.net/

  This file (cpp_bsaarchive.h) is part of the OpenMW package.

  OpenMW is distributed as free software: you can redistribute it
  and/or modify it under the terms of the GNU General Public License
  version 3, as published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  version 3 along with this program. If not, see
  http://www.gnu.org/licenses/ .

 */

#include <string>
#include <algorithm>

#ifndef BSA_BSA_ARCHIVE_H
#define BSA_BSA_ARCHIVE_H

namespace Bsa
{

/// Add the given BSA file as an input archive in the Ogre resource
/// system.
void addBSA(const std::string& file, const std::string& group="General");
void addDir(const std::string& file, const bool& fs, const std::string& group="General");

}

#endif
