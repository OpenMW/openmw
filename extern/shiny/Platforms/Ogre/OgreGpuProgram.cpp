#include <stdexcept>

#include "OgreGpuProgram.hpp"

#include <boost/lexical_cast.hpp>

#include <OgreHighLevelGpuProgramManager.h>
#include <OgreGpuProgramManager.h>
#include <OgreVector4.h>

namespace sh
{
	OgreGpuProgram::OgreGpuProgram(
		GpuProgramType type,
		const std::string& compileArguments,
		const std::string& name, const std::string& profile,
		const std::string& source, const std::string& lang,
		const std::string& resourceGroup)
		: GpuProgram()
	{
		Ogre::HighLevelGpuProgramManager& mgr = Ogre::HighLevelGpuProgramManager::getSingleton();
		assert (mgr.getByName(name).isNull() && "Vertex program already exists");

		Ogre::GpuProgramType t;
		if (type == GPT_Vertex)
			t = Ogre::GPT_VERTEX_PROGRAM;
		else
			t = Ogre::GPT_FRAGMENT_PROGRAM;

		mProgram = mgr.createProgram(name, resourceGroup, lang, t);
		if (lang != "glsl" && lang != "glsles")
			mProgram->setParameter("entry_point", "main");
		if (lang == "hlsl")
			mProgram->setParameter("target", profile);
		else if (lang == "cg")
			mProgram->setParameter("profiles", profile);

		mProgram->setSource(source);
		mProgram->load();

		if (mProgram.isNull() || !mProgram->isSupported())
			std::cerr << "Failed to compile shader \"" << name << "\". Consider the OGRE log for more information." << std::endl;
	}

	bool OgreGpuProgram::getSupported()
	{
		return (!mProgram.isNull() && mProgram->isSupported());
	}

	void OgreGpuProgram::setAutoConstant (const std::string& name, const std::string& autoConstantName, const std::string& extraInfo)
	{
		assert (!mProgram.isNull() && mProgram->isSupported());
		const Ogre::GpuProgramParameters::AutoConstantDefinition* d = Ogre::GpuProgramParameters::getAutoConstantDefinition(autoConstantName);

		if (!d)
			throw std::runtime_error ("can't find auto constant with name \"" + autoConstantName + "\"");
		Ogre::GpuProgramParameters::AutoConstantType t = d->acType;

		// this simplifies debugging for CG a lot.
		mProgram->getDefaultParameters()->setIgnoreMissingParams(true);

		if (d->dataType == Ogre::GpuProgramParameters::ACDT_NONE)
			mProgram->getDefaultParameters()->setNamedAutoConstant (name, t, 0);
		else if (d->dataType == Ogre::GpuProgramParameters::ACDT_INT)
			mProgram->getDefaultParameters()->setNamedAutoConstant (name, t, extraInfo == "" ? 0 : boost::lexical_cast<int>(extraInfo));
		else if (d->dataType == Ogre::GpuProgramParameters::ACDT_REAL)
			mProgram->getDefaultParameters()->setNamedAutoConstantReal (name, t, extraInfo == "" ? 0.f : boost::lexical_cast<float>(extraInfo));
	}
}
