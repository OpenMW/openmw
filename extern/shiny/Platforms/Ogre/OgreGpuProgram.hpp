#ifndef SH_OGREGPUPROGRAM_H
#define SH_OGREGPUPROGRAM_H

#include <string>

#include <OgreHighLevelGpuProgram.h>

#include "../../Main/Platform.hpp"

namespace sh
{
	class OgreGpuProgram : public GpuProgram
	{
	public:
		OgreGpuProgram (
			GpuProgramType type,
			const std::string& compileArguments,
			const std::string& name, const std::string& profile,
			const std::string& source, const std::string& lang,
			const std::string& resourceGroup);

		virtual bool getSupported();

		virtual void setAutoConstant (const std::string& name, const std::string& autoConstantName, const std::string& extraInfo = "");

	private:
		Ogre::HighLevelGpuProgramPtr mProgram;
	};
}

#endif
