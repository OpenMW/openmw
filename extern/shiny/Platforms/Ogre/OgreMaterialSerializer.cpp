#include "OgreMaterialSerializer.hpp"

#include <OgrePass.h>

#include <OgreStringConverter.h>

namespace sh
{
	void OgreMaterialSerializer::reset()
	{
		mScriptContext.section = Ogre::MSS_NONE;
		mScriptContext.material.setNull();
		mScriptContext.technique = 0;
		mScriptContext.pass = 0;
		mScriptContext.textureUnit = 0;
		mScriptContext.program.setNull();
		mScriptContext.lineNo = 0;
		mScriptContext.filename.clear();
		mScriptContext.techLev = -1;
		mScriptContext.passLev = -1;
		mScriptContext.stateLev = -1;
	}

	bool OgreMaterialSerializer::setPassProperty (const std::string& param, std::string value, Ogre::Pass* pass)
	{
		// workaround https://ogre3d.atlassian.net/browse/OGRE-158
		if (param == "transparent_sorting" && value == "force")
		{
			pass->setTransparentSortingForced(true);
			return true;
		}

		reset();

		mScriptContext.section = Ogre::MSS_PASS;
		mScriptContext.pass = pass;

		if (mPassAttribParsers.find (param) == mPassAttribParsers.end())
			return false;
		else
		{
			mPassAttribParsers.find(param)->second(value, mScriptContext);
			return true;
		}
	}

	bool OgreMaterialSerializer::setTextureUnitProperty (const std::string& param, std::string value, Ogre::TextureUnitState* t)
	{
		// quick access to automip setting, without having to use 'texture' which doesn't like spaces in filenames
		if (param == "num_mipmaps")
		{
			t->setNumMipmaps(Ogre::StringConverter::parseInt(value));
			return true;
		}
		
		reset();

		mScriptContext.section = Ogre::MSS_TEXTUREUNIT;
		mScriptContext.textureUnit = t;

		if (mTextureUnitAttribParsers.find (param) == mTextureUnitAttribParsers.end())
			return false;
		else
		{
			mTextureUnitAttribParsers.find(param)->second(value, mScriptContext);
			return true;
		}
	}

	bool OgreMaterialSerializer::setMaterialProperty (const std::string& param, std::string value, Ogre::MaterialPtr m)
	{
		reset();

		mScriptContext.section = Ogre::MSS_MATERIAL;
		mScriptContext.material = m;

		if (mMaterialAttribParsers.find (param) == mMaterialAttribParsers.end())
			return false;
		else
		{
			mMaterialAttribParsers.find(param)->second(value, mScriptContext);
			return true;
		}
	}
}
