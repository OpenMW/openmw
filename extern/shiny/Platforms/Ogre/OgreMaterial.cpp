#include "OgreMaterial.hpp"

#include <OgreMaterialManager.h>
#include <OgreTechnique.h>
#include <stdexcept>

#include "OgrePass.hpp"
#include "OgreMaterialSerializer.hpp"
#include "OgrePlatform.hpp"

namespace sh
{
	static const std::string sDefaultTechniqueName = "SH_DefaultTechnique";

	OgreMaterial::OgreMaterial (const std::string& name, const std::string& resourceGroup)
		: Material()
	{
		mName = name;
		assert (Ogre::MaterialManager::getSingleton().getByName(name).isNull() && "Material already exists");
		mMaterial = Ogre::MaterialManager::getSingleton().create (name, resourceGroup);
		mMaterial->removeAllTechniques();
		mMaterial->createTechnique()->setSchemeName (sDefaultTechniqueName);
		mMaterial->compile();
	}

	void OgreMaterial::ensureLoaded()
	{
		if (mMaterial.isNull())
			mMaterial = Ogre::MaterialManager::getSingleton().getByName(mName);
	}

	bool OgreMaterial::isUnreferenced()
	{
		// Resource system internals hold 3 shared pointers, we hold one, so usecount of 4 means unused
		return (!mMaterial.isNull() && mMaterial.useCount() <= Ogre::ResourceGroupManager::RESOURCE_SYSTEM_NUM_REFERENCE_COUNTS+1);
	}

	void OgreMaterial::unreferenceTextures()
	{
		mMaterial->unload();
	}

	OgreMaterial::~OgreMaterial()
	{
		if (!mMaterial.isNull())
			Ogre::MaterialManager::getSingleton().remove(mMaterial->getName());
	}

	boost::shared_ptr<Pass> OgreMaterial::createPass (const std::string& configuration, unsigned short lodIndex)
	{
		return boost::shared_ptr<Pass> (new OgrePass (this, configuration, lodIndex));
	}

	void OgreMaterial::removeAll ()
	{
		if (mMaterial.isNull())
			return;
		mMaterial->removeAllTechniques();
		mMaterial->createTechnique()->setSchemeName (sDefaultTechniqueName);
		mMaterial->compile();
	}

	void OgreMaterial::setLodLevels (const std::string& lodLevels)
	{
		OgreMaterialSerializer& s = OgrePlatform::getSerializer();

		s.setMaterialProperty ("lod_values", lodLevels, mMaterial);
	}

	bool OgreMaterial::createConfiguration (const std::string& name, unsigned short lodIndex)
	{
		for (int i=0; i<mMaterial->getNumTechniques(); ++i)
		{
			if (mMaterial->getTechnique(i)->getSchemeName() == name && mMaterial->getTechnique(i)->getLodIndex() == lodIndex)
				return false;
		}

		Ogre::Technique* t = mMaterial->createTechnique();
		t->setSchemeName (name);
		t->setLodIndex (lodIndex);
		if (mShadowCasterMaterial != "")
			t->setShadowCasterMaterial(mShadowCasterMaterial);

		mMaterial->compile();

		return true;
	}

	Ogre::MaterialPtr OgreMaterial::getOgreMaterial ()
	{
		return mMaterial;
	}

	Ogre::Technique* OgreMaterial::getOgreTechniqueForConfiguration (const std::string& configurationName, unsigned short lodIndex)
	{
		for (int i=0; i<mMaterial->getNumTechniques(); ++i)
		{
			if (mMaterial->getTechnique(i)->getSchemeName() == configurationName && mMaterial->getTechnique(i)->getLodIndex() == lodIndex)
			{
				return mMaterial->getTechnique(i);
			}
		}

		// Prepare and throw error message
		std::stringstream message;
		message << "Could not find configurationName '" << configurationName
				<< "' and lodIndex " << lodIndex;

		throw std::runtime_error(message.str());
	}

	void OgreMaterial::setShadowCasterMaterial (const std::string& name)
	{
		mShadowCasterMaterial = name;
		for (int i=0; i<mMaterial->getNumTechniques(); ++i)
		{
			mMaterial->getTechnique(i)->setShadowCasterMaterial(mShadowCasterMaterial);
		}
	}
}
