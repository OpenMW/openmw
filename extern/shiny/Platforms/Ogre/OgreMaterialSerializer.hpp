#ifndef SH_OGREMATERIALSERIALIZER_H
#define SH_OGREMATERIALSERIALIZER_H

#include <OgreMaterialSerializer.h>

namespace Ogre
{
	class Pass;
}

namespace sh
{
	/**
	 * @brief This class allows me to let Ogre handle the pass & texture unit properties
	 */
	class OgreMaterialSerializer : public Ogre::MaterialSerializer
	{
	public:
		bool setPassProperty (const std::string& param, std::string value, Ogre::Pass* pass);
		bool setTextureUnitProperty (const std::string& param, std::string value, Ogre::TextureUnitState* t);
		bool setMaterialProperty (const std::string& param, std::string value, Ogre::MaterialPtr m);

	private:
		void reset();
	};

}

#endif
