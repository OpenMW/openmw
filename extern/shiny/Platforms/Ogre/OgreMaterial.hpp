#ifndef SH_OGREMATERIAL_H
#define SH_OGREMATERIAL_H

#include <string>

#include <OgreMaterial.h>

#include "../../Main/Platform.hpp"

namespace sh
{
	class OgreMaterial : public Material
	{
	public:
		OgreMaterial (const std::string& name, const std::string& resourceGroup);
		virtual ~OgreMaterial();

		virtual boost::shared_ptr<Pass> createPass (const std::string& configuration, unsigned short lodIndex);
		virtual bool createConfiguration (const std::string& name, unsigned short lodIndex);

		virtual bool isUnreferenced();
		virtual void unreferenceTextures();
		virtual void ensureLoaded();

		virtual void removeAll ();

		Ogre::MaterialPtr getOgreMaterial();

		virtual void setLodLevels (const std::string& lodLevels);

		Ogre::Technique* getOgreTechniqueForConfiguration (const std::string& configurationName, unsigned short lodIndex = 0);

		virtual void setShadowCasterMaterial (const std::string& name);

	private:
		Ogre::MaterialPtr mMaterial;
		std::string mName;

		std::string mShadowCasterMaterial;
	};
}

#endif
