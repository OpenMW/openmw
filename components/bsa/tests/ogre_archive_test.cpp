#include <Ogre.h>
#include <iostream>

// This is a test of the BSA archive handler for OGRE.

#include "../bsa_archive.hpp"

using namespace Ogre;
using namespace std;

int main()
{
  // Disable Ogre logging
  new LogManager;
  Log *log = LogManager::getSingleton().createLog("");
  log->setDebugOutputEnabled(false);

  // Set up Root
  Root *root = new Root("","","");

  // Add the BSA
  Bsa::addBSA("../../data/Morrowind.bsa");

  // Pick a sample file
  String tex = "textures\\tx_natural_cavern_wall13.dds";
  cout << "Opening file: " << tex << endl;

  // Get it from the resource system
  DataStreamPtr data = ResourceGroupManager::getSingleton().openResource(tex, "General");

  cout << "Size: " << data->size() << endl;

  return 0;
}
