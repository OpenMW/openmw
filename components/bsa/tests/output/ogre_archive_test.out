Opening file: textures\tx_natural_cavern_wall13.dds
Size: 43808
